target/linux/ramips/dts/mt7621_beeline_smartbox-giga.dts

```diff
@@ -0,0 +1,237 @@
+	model = "Beeline SmartBox GIGA";
+
+	aliases {
+		label-mac-device = &gmac0;
+
+		led-boot = &led_status_green;
+		led-failsafe = &led_status_red;
+		led-running = &led_status_green;
+		led-upgrade = &led_status_red;
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		led_status_green: led-0 {
+			label = "green:status";
+			color = <LED_COLOR_ID_GREEN>;
+			function = LED_FUNCTION_STATUS;
+			gpios = <&gpio 13 GPIO_ACTIVE_HIGH>;
+		};
+
+		led-1 {
+			label = "blue:wan";
+			color = <LED_COLOR_ID_BLUE>;
+			function = LED_FUNCTION_WAN;
+			gpios = <&gpio 14 GPIO_ACTIVE_HIGH>;
+		};
+
+		led_status_red: led-2 {
+			label = "red:status";
+			color = <LED_COLOR_ID_RED>;
+			function = LED_FUNCTION_STATUS;
+			gpios = <&gpio 15 GPIO_ACTIVE_HIGH>;
+		};
+	};
+
+	keys {
+		compatible = "gpio-keys";
+
+		reset {
+			label = "reset";
+			gpios = <&gpio 4 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+		};
+	};
+
+	ubi-concat {
+		compatible = "mtd-concat";
+		devices = <&ubiconcat0 &ubiconcat1 &ubiconcat2 \
+			&ubiconcat3>;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "ubi";
+				reg = <0x0 0x5400000>;
+			};
+		};
+	};
+};
+
+&nand {
+	status = "okay";
+
+	partitions {
+		compatible = "sercomm,sc-partitions", "fixed-partitions";
+		#address-cells = <1>;
+		#size-cells = <1>;
+
+		partition@0 {
+			label = "u-boot";
+			reg = <0x0 0x100000>;
+			scpart-id = <0>;
+			read-only;
+		};
+
+		partition@100000 {
+			label = "dynamic partition map";
+			reg = <0x100000 0x100000>;
+			scpart-id = <1>;
+		};
+
+		factory: partition@200000 {
+			label = "Factory";
+			reg = <0x200000 0x100000>;
+			scpart-id = <2>;
+			read-only;
+
+			compatible = "nvmem-cells";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			macaddr_factory_21000: macaddr@21000 {
+				reg = <0x21000 0x6>;
+			};
+		};
+
+		partition@300000 {
+			label = "Boot Flag";
+			reg = <0x300000 0x100000>;
+			scpart-id = <3>;
+		};
+
+		partition@400000 {
+			label = "kernel";
+			reg = <0x400000 0x600000>;
+			scpart-id = <4>;
+		};
+
+		partition@a00000 {
+			label = "Kernel 2";
+			reg = <0xa00000 0x600000>;
+			scpart-id = <5>;
+			read-only;
+		};
+
+		ubiconcat0: partition@1000000 {
+			label = "ubiconcat0";
+			reg = <0x1000000 0x1800000>;
+			scpart-id = <6>;
+		};
+
+		partition@2800000 {
+			label = "File System 2";
+			reg = <0x2800000 0x1800000>;
+			scpart-id = <7>;
+			read-only;
+		};
+
+		ubiconcat1: partition@4000000 {
+			label = "ubiconcat1";
+			reg = <0x4000000 0x800000>;
+			scpart-id = <8>;
+		};
+
+		ubiconcat2: partition@4800000 {
+			label = "ubiconcat2";
+			reg = <0x4800000 0xc00000>;
+			scpart-id = <9>;
+		};
+
+		ubiconcat3: partition@5400000 {
+			label = "ubiconcat3";
+			reg = <0x5400000 0x2800000>;
+			scpart-id = <10>;
+		};
+
+		partition@7c00000 {
+			label = "bad block reserved";
+			reg = <0x7c00000 0x380000>;
+			scpart-id = <11>;
+			read-only;
+		};
```

Comment:
If I'm not mistaken, your Linux driver ignores all those offsets & sizes. It seems the are also dynamic (stored in struct on flash). What's the point of them in DT then?

Reply:
`compatible = "fixed-partitions"` is a fallback strategy if partition map is broken or missing. So, we need those offsets & sizes. And if a device has no bad blocks or bad blocks in some places all will be ok. 

IOW the layout differs from the reference one ("fixed-partitions") only if there are bad blocks on NAND. Partitions are shifted farther on the bad blocks size .

Partition map is ok:
```
Mon Mar 28 04:41:08 2022 kern.info kernel: [    1.096374] scpart: Valid 'SC PART MAP' (12 partitions) found at 0x100000
Mon Mar 28 04:41:08 2022 kern.notice kernel: [    1.109964] 12 scpart partitions found on MTD device mt7621-nand
Mon Mar 28 04:41:08 2022 kern.notice kernel: [    1.121917] Creating 12 MTD partitions on "mt7621-nand":
Mon Mar 28 04:41:08 2022 kern.notice kernel: [    1.132501] 0x000000000000-0x000000100000 : "u-boot"
Mon Mar 28 04:41:08 2022 kern.notice kernel: [    1.153524] 0x000000100000-0x000000200000 : "dynamic partition map"
Mon Mar 28 04:41:08 2022 kern.notice kernel: [    1.177207] 0x000000200000-0x000000300000 : "Factory"
Mon Mar 28 04:41:08 2022 kern.notice kernel: [    1.198437] 0x000000300000-0x000000400000 : "Boot Flag"
Mon Mar 28 04:41:08 2022 kern.notice kernel: [    1.219971] 0x000000400000-0x000000a00000 : "kernel"
Mon Mar 28 04:41:08 2022 kern.notice kernel: [    1.291609] 0x000000a00000-0x000001000000 : "Kernel 2"
Mon Mar 28 04:41:08 2022 kern.notice kernel: [    1.363526] 0x000001000000-0x000002800000 : "ubiconcat0"
Mon Mar 28 04:41:08 2022 kern.notice kernel: [    1.617160] 0x000002800000-0x000004000000 : "File System 2"
Mon Mar 28 04:41:08 2022 kern.notice kernel: [    1.871459] 0x000004000000-0x000004800000 : "ubiconcat1"
Mon Mar 28 04:41:08 2022 kern.notice kernel: [    1.963858] 0x000004800000-0x000005400000 : "ubiconcat2"
Mon Mar 28 04:41:08 2022 kern.notice kernel: [    2.096453] 0x000005400000-0x000007c00000 : "ubiconcat3"
Mon Mar 28 04:41:08 2022 kern.notice kernel: [    2.511879] 0x000007c00000-0x000007f80000 : "bad block reserved"
```
Partition map missing:
```
[    2.648797] scpart: No valid 'SC PART MAP' 
[    2.657036] 12 fixed-partitions partitions found on MTD device mt7621-nand 
[    2.670726] Creating 12 MTD partitions on "mt7621-nand": 
[    2.681318] 0x000000000000-0x000000100000 : "u-boot" 
[    2.702380] 0x000000100000-0x000000200000 : "dynamic partition map" 
[    2.726119] 0x000000200000-0x000000300000 : "Factory" 
[    2.747394] 0x000000300000-0x000000400000 : "Boot Flag" 
[    2.768901] 0x000000400000-0x000000a00000 : "kernel" 
[    2.840453] 0x000000a00000-0x000001000000 : "Kernel 2" 
[    2.912051] 0x000001000000-0x000002800000 : "ubiconcat0" 
[    3.164790] 0x000002800000-0x000004000000 : "File System 2" 
[    3.418140] 0x000004000000-0x000004800000 : "ubiconcat1" 
[    3.510168] 0x000004800000-0x000005400000 : "ubiconcat2" 
[    3.642439] 0x000005400000-0x000007c00000 : "ubiconcat3" 
[    4.056060] 0x000007c00000-0x000007f80000 : "bad block reserved"
```

Reply:
@rmilecki First of all, I would like to thank you for the review and useful feedback. I've started upstreaming before your review, so I need to send the version 2 after the changes. Please take a look, maybe there is something else that needs fixing.

@adschm @hauke @mans0n Can we at least assign a "new device" label, please?